### PR TITLE
Update proforma gem to 1.0

### PR DESCRIPTION
**Pull request by kkoehn**
Implement new fields from proforma, which were ignored before:
- submission_restrictions
- external_resources
- grading_hints

Todos

- [x] import/export implementation
- [x] add to show view

### CODE REVIEW

**Comment by kkoehn** [comment]
since we implemented a more generalized approach to handle `any_nodes`, it would make sense to handle the `meta_data` nodes that way too.
This PR is on hold, until this features has been implemented in proformaxml

**Comment by kkoehn** [comment]
I had to use the psql data-type `json` for those fields, because the order of keys is quite important for this use-case and:

> Because the json type stores an exact copy of the input text, it will preserve semantically-insignificant white space between tokens, as well as the order of keys within JSON objects. [docs](https://www.postgresql.org/docs/current/datatype-json.html)

Discussion: Do we need to switch the type for `meta_data` as well?
We might inadvertently modify the data some stores in this field.

**Comment by MrSerth** [comment]
> I had to use the psql data-type json for those fields, because the order of keys is quite important for this use-case.

Okay, I See. The order is, indeed, important, since it represents the order of nodes in the XML. This seems to be some limitation of the badgerfish standard. Potentially, I could think of some addition to our dachsfisch gem adding some order information, but this is currently not the case, obviously. Could you please add this fact (that the order is important) to the dachsfisch readme? Thanks.

Switching from `jsonb` to `json` has, besides keeping the order (and some white space) also some drawbacks on performance and the possibility to add indices. Therefore, I would be generally in favour of `jsonb`, but also see that this is currently not possible. This would mean that storing the dachsfisch data in PostgreSQL is just done as a "better string", validating the general JSON structure. Not ideal, but maybe that's something for ... later?

Besides my lowered preference for `json` (compared to `jsonb`), I am otherwise fine with that. 

> Discussion: Do we need to switch the type for meta_data as well?

Potentially, yes. Since this field also contains dachsfisch data (doesn't it?), the same limitations would apply. I think, as long as our dachsfisch output is not adding order information, we should maintain it manually by storing the data in a `json` column.

> I had to remove the `distinct:true` toggle for the ransack result ([f211b7e](https://github.com/openHPI/codeharbor/pull/990/commits/f211b7e8400faaa59d5fbe0ac849bf93d318a51a)), because postgres can't compare json fields (only jsonb). I'm not sure of the consequence of removing that toggle, but it may be wise to outsource the dachsfisch-node to another model. @MrSerth what do you think about this?

I am not quite sure about removing the `distinct: true` toggle. If I see this right, you were the one who added it with [this commit](https://github.com/openHPI/codeharbor/commit/7d1b74adfc97068a181714133ea3a0ea96f971df). The only use case I can think of are tasks visible for a group, since we are performing a join [there](https://github.com/openHPI/codeharbor/blob/46bc9560d62ec108f599a6cff7843988fd7f7585/app/models/task.rb#L43-L46). Have you tested how often a task appears in the list that has been added to two distinct groups? It could be duplicated...

**Review by MrSerth — COMMENTED** [review]
Nicely done, thank you! 👏 

I am pretty happy that those changes are really small, with the most changes being caused by our Dachsfisch changes (the introduction of the order). Otherwise, I mainly have a few questions about the migration, but nothing too big (I hope 😇).

Besides my questions there, I wasn't quite sure about the column defaults / null values of the corresponding columns:

- Why is the `meta_data` column mandatory? I have some cases where it just contains the default. I am just asking with a reference to the other three columns introduced, which are all optional.
- After running the migration, I still have some `meta_data` stored as `"{}"` (a string!), causing an empty meta data to be displayed in CodeHarbor.
- Can we change the migration not to update the `updated_at` timestamps? I think, this could be misleading, otherwise.

**Comment by kkoehn** [comment]
>     * Why is the `meta_data` column mandatory? I have some cases where it just contains the default. I am just asking with a reference to the other three columns introduced, which are all optional.

I made it mandatory, since it's mandatory in the [schema](https://github.com/ProFormA/proformaxml/blob/master/proforma.xsd#L628). The other three new columns (+  test_meta_data) are optional. I wanted to reflect that in the schema.
 
>     * After running the migration, I still have some `meta_data` stored as `"{}"` (a string!), causing an empty meta data to be displayed in CodeHarbor.

Everything that was empty before should now have `{}`, I will change the default from string to hash in the migration.

>     * Can we change the migration not to update the `updated_at` timestamps? I think, this could be misleading, otherwise.

sure

**Comment by MrSerth** [comment]
> > ```
> > * Why is the `meta_data` column mandatory? I have some cases where it just contains the default. I am just asking with a reference to the other three columns introduced, which are all optional.
> > ```
> 
> I made it mandatory, since it's mandatory in the [schema](https://github.com/ProFormA/proformaxml/blob/master/proforma.xsd#L628). The other three new columns (+ test_meta_data) are optional. I wanted to reflect that in the schema.

Ah, that makes sense, of course. Thanks!

> 
> > ```
> > * After running the migration, I still have some `meta_data` stored as `"{}"` (a string!), causing an empty meta data to be displayed in CodeHarbor.
> > ```
> 
> Everything that was empty before should now have `{}`, I will change the default from string to hash in the migration.

Perfect 👌 

> 
> > ```
> > * Can we change the migration not to update the `updated_at` timestamps? I think, this could be misleading, otherwise.
> > ```
> 
> sure

👍

File: db/migrate/20231002220433_migrate_meta_data_to_dachsfisch.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+class MigrateMetaDataToDachsfisch < ActiveRecord::Migration[7.0]
+  def change
+    Task.all.filter {|t| !t.meta_data.nil? }.each {|task| task.update(meta_data: transform_meta_data(task.meta_data)) }
+    Test.all.filter {|t| !t.meta_data.nil? }.each {|test| test.update(meta_data: transform_meta_data(test.meta_data)) }
```
[the remark this answers is not in the document]
Comment (kkoehn) [reply db/migrate/20231002220433_migrate_meta_data_to_dachsfisch.rb]: I don't see a problem there, what kind of problem do you expect?

[the remark this answers is not in the document]
Comment (MrSerth) [reply db/migrate/20231002220433_migrate_meta_data_to_dachsfisch.rb]: I wasn't sure whether the missing call to `deep_symbolize_keys` has some undesired effect. When reading the code, I would think that it's fine, so let's continue :)

**Review by MrSerth — APPROVED** [review]
Very nice work; it's so great to finally see your hard work of the past months yielding in that awesome and clean PR; finally enabling full compatibility with the ProFormA standard. Many thanks for taking the time and energy to dig through all the issues you've encountered along the way -- I would say that the results are paying off! 👏

**Comment by MrSerth** [comment]
I was just rebasing on the latest master branch (without any code changes) and will now proceed merging.

**Pull request merged**